macOS: Import Chrome new tab shortcuts as favorites

rachelmcr opened the pull request:
Task/Issue URL: https://app.asana.com/1/137249556945/project/1209825025475019/task/1210706281466357?focus=true

### Description

This PR updates the bookmarks import feature to import Chrome new tab shortcuts as favorites in the new tab page. This new behavior is behind a feature flag and is disabled by default.

#### Implementation

* We have a new local feature flag `updatedBookmarksFavoritesImport` to control the new behavior.
* We read the shortcuts settings from Chrome preferences to determine how the Chrome new tab page is set up.
* If Chrome shortcuts are enabled, we fetch the shortcuts (from preferences for custom shortcuts or from the Top Sites database for the auto-generated “most visited” shortcuts) and merge them with the Chrome bookmarks.
* When the bookmarks are imported, we now collect the favorited bookmarks and add them to favorites all at once. This allows us to preserve the order of the Chrome shortcuts (while retaining the same ordering for other source browsers).

### Testing Steps

#### New behavior (feature flag enabled)

Set up Chrome:
1. Create a new Chrome profile.
2. Add at least one bookmark (e.g. visit a site and tap the star to bookmark it).
3. Choose one of the following setup options:
   * Visit several sites and confirm they appear on the New Tab Page, but don’t edit the shortcuts list.
   * Open the New Tab Page and add some shortcuts using the + button or edit the shortcuts list using the “…” menu on each shortcut.
   * Tap “Customize Chrome” in the bottom right corner and disable the “Show shortcuts” setting.
   * Tap “Customize Chrome” in the bottom right corner and select the “Most-visited sites” option.
4. Follow the steps below to test the import.

Test the import:
1. Go to Debug menu > Feature flag overrides > Other > enable `updatedBookmarksFavoritesImport` flag.
2. Go to Bookmarks menu > Import bookmarks.
3. Follow the prompts in the UI to import Chrome bookmarks.
4. Confirm your new tab Favorites section shows the same items as the Chrome new tab shortcuts.

Additional scenarios:
* Import twice and confirm your bookmarks are imported a second time but you don’t have duplicate favorites.
* Import bookmarks from a non-Chromium browser and confirm they work as before.

#### Previous behavior (feature flag disabled)

Set up Chrome:
1. Create a new Chrome profile.
2. Add a least one bookmark in the bookmarks bar: Visit a site, tap the star, and select “Bookmarks Bar” as the folder.
3. Add at least one bookmark in another folder: Visit a site, tap the star, and select “Other Bookmarks” as the folder.
4. Follow the steps below to test the import.

Test the import:
1. Go to Debug menu > Feature flag overrides > Other > disable `updatedBookmarksFavoritesImport` flag.
2. Go to Bookmarks menu > Import bookmarks.
5. Follow the prompts in the UI to import Chrome bookmarks.
6. Confirm your new tab Favorites section shows the same items as the Chrome bookmarks bar.

### Impact and Risks
Medium: Could disrupt specific features or user flows

#### What could go wrong?
The following scenarios have been testing manually and covered with unit tests:
* Shortcuts are imported and/or bookmarks bar is not imported as favorites when feature flag is disabled
* Shortcuts are not imported and/or bookmarks bar is imported as favorites when feature flag is enabled
* Shortcuts are merged incorrectly with bookmarks or added in wrong order to favorites

### Quality Considerations
* We plan to add success/failure pixels for importing these favorites (under review)
* If a Chromium browser does not implement the new tab shortcuts this way, they won’t be imported.

### Notes to Reviewer

Chrome shortcuts|Imported favorites
-|-
![Screenshot 2025-07-03 at 17 39 35](https://github.com/user-attachments/assets/6b484de0-d17f-407b-8bb3-39d3f40c3376)|![Screenshot 2025-07-03 at 17 39 47](https://github.com/user-attachments/assets/3d163695-fe8d-4f56-b627-276712645b5c)


—
###### Internal references:
[Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f) | [Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552) | [Tech Design Template](https://app.asana.com/0/59792373528535/184709971311943)

rachelmcr commented on macOS/DuckDuckGo/DataImport/ChromiumPreferences.swift:
```diff
@@ -33,6 +33,20 @@ struct ChromiumPreferences: Decodable {
     }
 
+    struct NewTabPage: Decodable {
+        let shortcustVisible: Bool?
+        let useMostVisitedTiles: Bool?
```
The typo is in the Chrome preferences key itself (`shortcust_visible`). I kept the typo to match the key (rather than adding custom coding keys) but we can access the value with the new property `isNewTabShortcutsEnabled`.

mallexxx replied:
the comments should be added to the code

rachelmcr commented on macOS/UnitTests/DataImport/FavoritesUtilsTests.swift:
```diff
@@ -0,0 +1,77 @@
+
+struct FavoritesUtilsTests {
+
+    @Test func testMergeBookmarksAndFavorites_ReturnsBookmarks_WhenFavoritesEmpty() {
```
Ah that’s right, thanks. Updated those tests in dfd7473.

mallexxx commented on macOS/DuckDuckGo/DataImport/Logins/Chromium/ChromiumDataImporter.swift:
```diff
@@ -30,12 +31,14 @@ internal class ChromiumDataImporter: DataImporter {
+    private let featureFlagger: FeatureFlagger
 
-    init(profile: DataImport.BrowserProfile, loginImporter: LoginImporter?, bookmarkImporter: BookmarkImporter, faviconManager: FaviconManagement) {
+    init(profile: DataImport.BrowserProfile, loginImporter: LoginImporter?, bookmarkImporter: BookmarkImporter, faviconManager: FaviconManagement, featureFlagger: FeatureFlagger = Application.appDelegate.featureFlagger) {
```
Please remove the default parameter and pass the `Application.appDelegate.featureFlagger` from `DataImportViewModel`

mallexxx commented on macOS/DuckDuckGo/DataImport/Logins/Chromium/ChromiumDataImporter.swift:
```diff
@@ -147,6 +155,36 @@ internal class ChromiumDataImporter: DataImporter {
+                }
+            }
+        } catch {
+            return []
```
Should we consider firing an error Pixel in case the Top site reading fails?
Otherwise we‘re loosing all the error info carefully collected in the ChromiumTopSitesReader

rachelmcr replied:
Oh yeah, that’s the plan — I’ll add comments about the planned pixels.

mallexxx commented on macOS/DuckDuckGo/Bookmarks/Services/LocalBookmarkStore.swift:
```diff
@@ -798,25 +803,24 @@ final class LocalBookmarkStore: BookmarkStore {
-                                                   in: context)
-
+            let (result, favorites) = recursivelyCreateEntitiesAndCollectFavorites(from: children,
+                                                                                    parent: folderParent,
+                                                                                    in: context)
```
indentation

mallexxx commented on macOS/DuckDuckGo/DataImport/Bookmarks/FavoritesUtils.swift:
```diff
@@ -0,0 +1,96 @@
+        func processBookmarkOrFolder(_ item: ImportedBookmarks.BookmarkOrFolder) -> ImportedBookmarks.BookmarkOrFolder {
+            var updatedItem = item
+
+            if !item.isFolder, let shortcut = favorites.first(where: { $0.urlString == item.urlString }) {
```
should we instead make a `Set(favorites.map(\.urlString))` for faster further checks?

mallexxx commented on macOS/DuckDuckGo/DataImport/Logins/Chromium/ChromiumDataImporter.swift:
```diff
@@ -105,7 +107,13 @@ internal class ChromiumDataImporter: DataImporter {
+                }
             }
 
             if case .success = bookmarksSummary {
```
number of bookmarks reported by 
```
updateProgress(.importingBookmarks(numberOfBookmarks: try? bookmarkResult.get().numberOfBookmarks
```
wouldn't match actual number of imported bookmarks in case when we append new bookmarks from shortcuts in `mergeBookmarksAndFavorites`

mallexxx commented on macOS/DuckDuckGo/DataImport/Bookmarks/FavoritesUtils.swift:
```diff
@@ -0,0 +1,96 @@
+//  limitations under the License.
+//
+
+final class FavoritesUtils {
```
it should be `enum` (not `class` or `struct`) as it shouldn‘t be instantiated

mallexxx commented on macOS/DuckDuckGo/DataImport/Bookmarks/FavoritesUtils.swift:
```diff
@@ -0,0 +1,96 @@
+            return updatedItem
+        }
+
+        let updatedBookmarkBar = bookmarks.topLevelFolders.bookmarkBar.map(processBookmarkOrFolder)
+        let updatedOtherBookmarks = bookmarks.topLevelFolders.otherBookmarks.map(processBookmarkOrFolder)
+        let updatedSyncedBookmarks = bookmarks.topLevelFolders.syncedBookmarks.map(processBookmarkOrFolder)
```
Given some users may have thouthands of bookmarks and the `processBookmarkOrFolder` not actually performing deletions/insertions but only object mutations, can the triple recursive pass impose memory overusage issues?
Probably the recursion itself shouldn‘t be a big deal for common bookmarks folder structure, but I‘d reconsider converting this into in-place array mutation like:
```suggestion
static func mergeBookmarksAndFavorites(bookmarks: inout ImportedBookmarks, favorites: [ImportedBookmarks.BookmarkOrFolder]) {
        guard !favorites.isEmpty else { return }

        let favoriteUrlIndices = favorites.reduce(into: [String: Int]()) { result, favorite in
            guard let urlString = favorite.urlString else { return }
            result[urlString] = favorite.favoritesIndex
        }
        var foundFavoriteUrls: Set<String> = []

        // Recursively process bookmarks to mark matching URLs as favorites
        func processMaybeFolder(_ folder: inout ImportedBookmarks.BookmarkOrFolder?) {
            if folder != nil {
                processFolder(&folder!)
            }
        }
        func processFolder(_ folder: inout ImportedBookmarks.BookmarkOrFolder) {
            assert(folder.isFolder, "Folder must be passed")
            if folder.children != nil {
                processFolderChildren(&folder.children!)
            }
        }
        func processFolderChildren(_ items: inout [ImportedBookmarks.BookmarkOrFolder]) {
            for idx in items.indices {
                guard !items[idx].isFolder else {
                    processFolder(&items[idx])
                    continue
                }

                guard let urlString = items[idx].urlString else { continue }

                if let favoriteIndex = favoriteUrlIndices[urlString] {
                    items[idx].isDDGFavorite = true
                    items[idx].favoritesIndex = favoriteIndex

                    foundFavoriteUrls.insert(urlString)
                }
            }
        }

        processMaybeFolder(&bookmarks.topLevelFolders.bookmarkBar)
        processMaybeFolder(&bookmarks.topLevelFolders.otherBookmarks)
        processMaybeFolder(&bookmarks.topLevelFolders.syncedBookmarks)

        // Create bookmarks objects for favorite shortcuts that don't have matching bookmarks and add them to bookmark bar
        let uniqueShortcuts = favorites.filter { !foundFavoriteUrls.contains($0.urlString ?? "") }

        bookmarks.topLevelFolders.bookmarkBar?.children?.append(contentsOf: uniqueShortcuts) ?? {
            bookmarks.topLevelFolders.bookmarkBar = .folder(name: bookmarks.topLevelFolders.bookmarkBar?.name ?? "", children: uniqueShortcuts)
        }()
    }
```
this would require turning `BookmarkOrFolder.children`, `TopLevelFolders`‘s properties and `ImportedBookmarks.topLevelFolders` into `var` but it shouldn‘t be a problem as it‘s a `struct`

Additionally, this optimizing change would be required to apply to `ChromiumDataImporter`:
```suggestion
let bookmarkResult = bookmarkReader.readBookmarks()

guard case .success(var importedBookmarks) = bookmarkResult else {
    summary[.bookmarks] = .failure(bookmarkResult.error!)
    return summary
}

try updateProgress(.importingBookmarks(numberOfBookmarks: importedBookmarks.numberOfBookmarks,
                                       fraction: passwordsFraction + dataTypeFraction * 0.5))

var markRootBookmarksAsFavoritesByDefault = true
if featureFlagger.isFeatureOn(.updatedBookmarksFavoritesImport) {
    markRootBookmarksAsFavoritesByDefault = false
    let newTabShortcuts = fetchShortcutsAsFavorites()

    FavoritesUtils.mergeBookmarksAndFavorites(bookmarks: &importedBookmarks, favorites: newTabShortcuts)
}

let bookmarksSummary = bookmarkImporter.importBookmarks(importedBookmarks,
                                                        source: .thirdPartyBrowser(source),
                                                        markRootBookmarksAsFavoritesByDefault: markRootBookmarksAsFavoritesByDefault,
                                                        maxFavoritesCount: nil)
summary[.bookmarks] = .success(.init(bookmarksSummary))

await importFavicons()

try updateProgress(.importingBookmarks(numberOfBookmarks: bookmarksSummary.successful,
                                       fraction: passwordsFraction + dataTypeFraction * 1.0))
```

mallexxx replied:
I‘m open for discussion on the suggestion change, but for me it looks better readable and more effective

mallexxx commented on macOS/DuckDuckGo/DataImport/ChromiumPreferences.swift:
```diff
@@ -77,4 +94,23 @@ struct ChromiumPreferences: Decodable {
+        ntp?.shortcustVisible ?? false
+    }
+
+    /// Returns `true` if the New Tab Page is using auto-generated shortcuts.
+    /// Returns `false` if the New Tab Page is using custom shortcuts.
+    var isUsingAutoGeneratedShortcuts: Bool {
```
not a big deal, but if it requires a comment to describe what is the boolean variable about, maybe consider introducing a enum in place of it:
```
enum NewTabPageShortcutsStyle {
  case autoGenerated, custom
}
var newTabPageShortcutsStyle: NewTabPageShortcutsStyle {
  (ntp?.useMostVisitedTiles ?? false) ? .autoGenerated : .custom
}
```

mallexxx reviewed (changes requested):
The change looks good overall 👍
Please consider taking in my suggestion to increase efficiency.
Unfortunately, the testing steps didn‘t work for me. What I did:
- Enabled the feature flag
- Created a new profile in Chrome
- Opened the New Tab Page, added couple of shortcuts using the + button
- in DDG: Import bookmarks, Chrome -> new profile (although it displays the correct profile name only after editing the profile in Chrome and changing its name/picture)
- Import bookmarks -> Error (no Bookmarks file found)
- then I‘ve slightly modified `ChromiumDataImporter` to allow it to proceed with Shortcuts import even after the `bookmarkResult` is a `.failure`: then `preferences.isNewTabShortcutsEnabled` was `false` and didn‘t proceed with shortcuts import
- I‘ve commented out `preferences.isNewTabShortcutsEnabled` - even then the import resulted in 0 bookmarks.
- Then I replaced `preferences.isUsingAutoGeneratedShortcuts` with `true` for the condition to always pass: it did import only the default "Web Store" shortcut
- Then I‘ve opened all the newly created shortcuts in Chrome - only then the import of all the item did succeed.

my Chrome is v. 138.0.7204.101
Also the Chrome setup steps are unclear, could you please update it to provide the exact step-by-step instructions

mallexxx commented on macOS/DuckDuckGo/DataImport/Bookmarks/FavoritesUtils.swift:
```diff
@@ -0,0 +1,96 @@
+
+    /// Merges imported bookmarks with the provided array of favorites, marking bookmarks with matching URLs as favorites.
+    /// Any favorites not matching existing bookmarks will be added to the bookmark bar.
+    static func mergeBookmarksAndFavorites(bookmarks: ImportedBookmarks, favorites: [ImportedBookmarks.BookmarkOrFolder]) -> ImportedBookmarks {
```
ok, please consider find it some better naming, currently it‘s hard to reason about what‘s "FavoritesUtils" is about, it may be misleading whether it‘s related to our database storage or favorites view…

rachelmcr commented:
Thanks for the review! I really appreciate your suggestions and I’ve made the suggested changes.

The issues you had with the testing steps are related to default Chrome behavior that I originally missed. I’ve reviewed all the setup scenarios I’m aware of, updated the behavior to handle them (in 59bd322), and updated the testing steps with more explicit instructions.

I also added a small improvement to how favorites are de-duped in bb32d1e (comparing `naked` URLs, not just identical URLs).

jotaemepereira commented on macOS/DuckDuckGo/Bookmarks/Services/LocalBookmarkStore.swift:
```diff
@@ -715,7 +715,7 @@ final class LocalBookmarkStore: BookmarkStore {
     /// 3. **Chrome:** Put all bookmarks at the root level, except for Other Bookmarks which go in a root level "Other Bookmarks" folder.
     /// 4. **Firefox:** Put all bookmarks at the root level, except for Other Bookmarks which go in a root level "Other Bookmarks" folder.
-    func importBookmarks(_ bookmarks: ImportedBookmarks, source: BookmarkImportSource) -> BookmarksImportSummary {
+    func importBookmarks(_ bookmarks: ImportedBookmarks, source: BookmarkImportSource, markRootBookmarksAsFavoritesByDefault: Bool = true) -> BookmarksImportSummary {
```
💡 Can we modify the function comment to explain what’s the usage of `markRootBookmarksAsFavoritesByDefault`?

jotaemepereira commented on macOS/DuckDuckGo/DataImport/Bookmarks/FavoritesImportProcessor.swift:
```diff
@@ -0,0 +1,70 @@
+        guard !favorites.isEmpty else { return }
+
+        let favoriteUrlIndices = favorites.reduce(into: [String: Int]()) { result, favorite in
+            guard let urlString = favorite.url?.nakedString else { return }
```
💡 There is an edge case here where we can improve the performance. If there is a list of favorites but the the `url` is incorrect for all them, we still go through all the bookmarks recurisvely. 

We should return earlier if the list of `favoriteUrlIndices` is empty.

We can just do `guard !favoriteUrlIndices.isEmpty else { return }` after this is calculated.

jotaemepereira reviewed (approved):
@rachelmcr This looks great, thanks for the add-on and the extensive testing. I’ve added two comments. 

I know that you still need to add pixels, but how you will know the error is because of this new add-on (let’s say Chrome changes the structure)? 

It is important for the future that we know if this part of the import fails, so we can use the FF to turn it off.
